File: Java/basic/DifferenceHighLow.java
/*
 * This is free to use as it was only made for practice.
 */

package com.jhazelle.gfg.basic;

import java.util.HashMap;
import java.util.Scanner;

/**
 * This is a solution to the <a href="http://practice.geeksforgeeks.org/problems/difference-between-highest-and-lowest-occurrence/0" target="_blank">GeeksForGeeks Basic Difference between highest and lowest</a>
 * 
 * To summarize the practice:
 * Given an array, find the difference between the highest and lowest numbers in the array.
 * Print the difference.
 * 
 * @author <a href="mailto:jhazelle@asu.edu">Justin Hazelle</a>
 * <a href="https://github.com/BinaryWrought" target="_blank">GitHub</a>
 */
public class DifferenceHighLow 
{

    /**
     * @param args the command line arguments
     */
    public static void main( String[] args ) 
	{
		Scanner scanner = new Scanner( System.in );													//create a scanner object to get user input
		try																							//this try block is to catch the inevitable exception caused by the user quitting 
		{
			int t = scanner.nextInt();																//get number of test cases
			
			while( t > 0 )																			//So long as there are more test cases
			{	
				int n = scanner.nextInt();															//get the size of the input array
				HashMap< Integer, Integer > map = new HashMap<>();									//create a map
				for( int i = 0; i < n; ++i )														//loop to populate array
				{
					int key = scanner.nextInt();													//get the input
					if( map.containsKey( key ) )													//check if the map already has the key
					{
						int val = map.get( key ) + 1;												//if so, get the current value
						map.put( key, val );														//and update the key
					}
					else
						map.put( key, 1 );															//otherwise, add the new key with initial value 1
				}
				
				int max = -1;																		//container for most frequent occurance
				int min = 111111;																	//container for least frequent occurance
				
				for( Integer i : map.values() )														//iterate through the map
				{
					if( i > max )																	//look for the largest occurrance
						max = i;																	//set it to the container
					if( i < min )																	//look for the smallest occurrance
						min = i;																	//set it to the container
				}
				
				System.out.println( ( max - min ) );												//print the difference
				
				t--;																				//get next test case		
			}
		}
		catch( Exception e )																		//catch an exception thrown by incorrect input
		{
			
		}
	}
}

File: Python/basic/DifferenceHighLow.py
#
# This is a solution to the <a href="http://practice.geeksforgeeks.org/problems/difference-between-highest-and-lowest-occurrence/0" target="_blank">GeeksForGeeks Basic Difference between highest and lowest</a>
#
# To summarize the practice:
# Given an array, find the difference in occurrances between the highest and lowest numbers in the array.
# Print the difference.
#
# @author <a href="mailto:jhazelle@asu.edu">Justin Hazelle</a>
# <a href="https://github.com/BinaryWrought" target="_blank">GitHub</a>

t = int( input() )                          #get number of test cases
for _ in range( t ):                        #loop for each test case
    n = int(input())                        #the number of integers in the list
    a = list(map(int, input("").split()))   #create a list of ints mapped from the input
    s = set(a)                              #create a set from the list
    least = n                               #initialize to n because the least occurring number will be less than n
    most = 0                                #initialize to 0 becuase the most occurring number will be more than 0
    for i in s:                             #loop through the unique elements of the list
        c = a.count(i)                      #see how many times this element appears in the list
        if c < least:                       #if it appears less than the current least amount...
            least = c                       #set it to the current least amount
        if c > most:                        #if it appears more than the current most amount...
            most = c                        #set it to the current most amount
    print( most - least )                   #print the difference between the most occurring number and the least
    